### PR TITLE
Implemented tests for all services

### PR DESCRIPTION
**Pull request by holsoni**
Here is coverage amount 
image

### CODE REVIEW

**Review by AntonZhdanov — APPROVED** [review]

**Comment by AntonZhdanov** [comment]
Good solution.

**Review by OlegTuryk — APPROVED** [review]

File: src/test/java/core/basesyntax/service/impl/DataProcessingServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    @BeforeEach
+    void setUp() {
+        transactionStrategy = new TransactionStrategyImpl();
+        dataProcessingService = new DataProcessingServiceImpl(transactionStrategy);
+    }
+
+    @Test
+    void processData_OK() {
+        List<FruitTransaction> transactions = List.of(
+                new FruitTransaction(FruitShopOperation.BALANCE, "apple", 40),
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/impl/DataProcessingServiceImplTest.java:28]: image

You can test only 1 operation here to verify if you call handler correctly, you don't need to test logic of strategy or handlers here.

File: src/test/java/core/basesyntax/service/impl/TransactionValidatorImplTest.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    void setUp() {
+        validator = new TransactionValidatorImpl();
+    }
+
+    @Test
+    void validate_OK() {
+        String input = HEADING + System.lineSeparator()
+                + VALID_RECORD + System.lineSeparator();
+        validator.validate(input);
+        Assertions.assertDoesNotThrow(() -> new InvalidDataException("Data is invalid"));
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/impl/TransactionValidatorImplTest.java:33]: Are you sure that you are using this assertion correctly? You should pass runnable with executable, not exception, similar to what you did in `assertThrows`

File: src/test/java/core/basesyntax/service/impl/TransactionsParserImplTest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import core.basesyntax.model.FruitShopOperation;
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.interfaces.TransactionParser;
+import java.util.List;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class TransactionsParserImplTest {
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/impl/TransactionsParserImplTest.java:11]: Add test for edge case.

File: src/test/java/core/basesyntax/service/transaction/impl/BalanceTransactionHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import core.basesyntax.model.FruitShopOperation;
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.transaction.TransactionHandler;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class BalanceTransactionHandlerTest {
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/transaction/impl/BalanceTransactionHandlerTest.java:13]: Add test for edge case.

File: src/test/java/core/basesyntax/service/transaction/impl/BalanceTransactionHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                FruitShopOperation.BALANCE, "apple", 60);
+        FruitTransaction transactionApple = new FruitTransaction(
+                FruitShopOperation.BALANCE, "banana", 60);
+        Map<String, Integer> expected = new HashMap<>(
+                Map.of("apple", 60, "banana", 60));
+        balanceTransactionHandler.executeTransaction(transactionPineapple);
+        balanceTransactionHandler.executeTransaction(transactionApple);
+        Assertions.assertEquals(expected, Storage.getAll(),
+                "Balance operation doesn't work correctly: ");
+        Storage.clear();
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/transaction/impl/BalanceTransactionHandlerTest.java:33]: Clear storage in BeforeEach or AfterEach.

File: src/test/java/core/basesyntax/service/transaction/impl/PurchaseTransactionHandlerTest.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    @Test
+    void executePurchaseTransaction_OK() {
+        FruitTransaction transaction = new FruitTransaction(
+                FruitShopOperation.PURCHASE, "apple", 60);
+        Map<String, Integer> expected = new HashMap<>(
+                Map.of("apple", 0,"banana", 20));
+        purchaseTransactionHandler.executeTransaction(transaction);
+        Assertions.assertEquals(expected, Storage.getAll(),
+                "Purchase operation doesn't work correctly: ");
+        Storage.clear();
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/transaction/impl/PurchaseTransactionHandlerTest.java:33]: Clear storage in BeforeEach or AfterEach.

File: src/test/java/core/basesyntax/service/transaction/impl/PurchaseTransactionHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    void executeReturnTransaction_NotOk() {
+        FruitTransaction transaction = new FruitTransaction(
+                FruitShopOperation.PURCHASE, "apple", 80);
+        InvalidDataException actual = Assertions.assertThrows(InvalidDataException.class, () -> {
+            purchaseTransactionHandler.executeTransaction(transaction);
+        });
+        String expectedMessage = "The amount of purchase is too big! "
+                + "Purchase amount is 80, but shop has only 60 apple.";
+        Assertions.assertEquals(expectedMessage, actual.getMessage());
+        Storage.clear();
```
Comment (vovasalyha) [inline src/test/java/core/basesyntax/service/transaction/impl/PurchaseTransactionHandlerTest.java:46]: Clear storage in BeforeEach or AfterEach.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Good solution. Let's improve it a bit

File: src/test/java/core/basesyntax/service/impl/TransactionStrategyImplTest.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+                TransactionStrategyImpl.handlers.get(FruitShopOperation.RETURN);
+        Assertions.assertEquals(returnHandler,
+                transactionStrategy.get(FruitShopOperation.RETURN));
+    }
+
+    @Test
+    void getSupplyHandler_OK() {
+        TransactionHandler supplyHandler =
+                TransactionStrategyImpl.handlers.get(FruitShopOperation.SUPPLY);
+        Assertions.assertEquals(supplyHandler,
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/impl/TransactionStrategyImplTest.java:30]: this assert is kind of useless for me, because it similar to: assertEquals(actual, actual) is it true always, better to make assertEquals(SomeHandles.class, actual.getClass)

File: src/test/java/core/basesyntax/service/impl/TransactionValidatorImplTest.java (hunk 2 above)
Comment (andrii-hoienko) [reply src/test/java/core/basesyntax/service/impl/TransactionValidatorImplTest.java]: Not resolved

File: src/test/java/core/basesyntax/service/transaction/impl/BalanceTransactionHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
```
Comment (andrii-hoienko) [inline src/test/java/core/basesyntax/service/transaction/impl/BalanceTransactionHandlerTest.java:34]: Please remove this, it already happens in setUp (see all your tests you have to remove it everywhere when this operation done in beforeEach/afterEach)

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, but some improvements required
